invenio_drafts_resources/auditlog_actions.py

```diff
@@ -8,30 +8,119 @@
-            message_template="User {user_id} created the draft {resource_id}.",
-        ),
-        "draft.edit": AuditAction(
-            name="draft.edit",
-            message_template="User {user_id} updated the draft {resource_id}.",
-        ),
-        "record.publish": AuditAction(
-            name="record.publish",
-            message_template="User {user_id} published the record {resource_id}.",
-        ),
-        "draft.delete": AuditAction(
-            name="draft.delete",
-            message_template="User {user_id} deleted the draft {resource_id}.",
-        ),
-        "draft.new_version": AuditAction(
-            name="draft.new_version",
-            message_template="User {user_id} created a new version {resource_id}.",
-        ),
-    }
+from invenio_access.permissions import system_user_id
+from invenio_audit_logs.services import AuditLogAction
+from invenio_records.dictutils import dict_lookup, dict_set
+from invenio_users_resources.entity_resolvers import UserResolver
+
+
+class UserContext:
+    """Payload generator for audit log using the user entity resolver."""
+
+    def __init__(self, key="user"):
+        """Ctor."""
+        self.key = key
+        self.resolver = UserResolver()
+
+    def __call__(self, data, lookup_key="user_id", **kwargs):
+        """Update data with resolved user data."""
+        entity_ref = dict_lookup(data, lookup_key)
+        entity_proxy = self.resolver.get_entity_proxy({self.key: entity_ref})
+        if entity_ref == system_user_id:
+            entity_data = entity_proxy.system_record()
+            entity_data = {
+                "id": str(entity_data["id"]),
+                "name": str(entity_data["username"]),
+                "email": str(entity_data["email"]),
+            }
+        else:
+            entity_data = entity_proxy.resolve()
+            entity_data = {
+                "id": str(entity_data.id),
+                "name": entity_data.username,
+                "email": entity_data.email,
+            }
+        dict_set(data, self.key, entity_data)
+
+
+class RecordContext:
+    """Payload generator for audit log to get record auditing metadata."""
+
+    def __init__(self, revision_id=False, parent_pid=False):
+        """Ctor."""
+        # Enable or Disable adding metadata
+        self.revision_id = revision_id
+        self.parent_pid = parent_pid
+
+    def __call__(self, data, **kwargs):
+        """Update data with resolved record data."""
+        record = kwargs.get("record", None)
+        if record is None:
+            return
+        if self.revision_id:
+            record_versions = record.model.versions.all()
+            dict_set(data, "metadata.revision_id", record_versions[-1].transaction_id)
+        if self.parent_pid:
+            dict_set(data, "metadata.parent_pid", record.parent.pid.pid_value)
+
+
+class RecordBaseAuditLog(AuditLogAction):
+    """Base class for audit log builders."""
+
+    context = [
+        UserContext(),
```

Comment:
@utnapischtim to answer your question in https://github.com/inveniosoftware/invenio-drafts-resources/pull/305#discussion_r2098406467
we have a problem with circular dependency at the moment, because of user context. Normally I would say this class should go to audit logs ... but if we want to use audit logs to log action on users? same dependency issue will come up if we implement this class in records-resources. 
We didn't find a good place yet.

Reply:
yes, maybe the circular dependency is not fixed, but other functionalities which are using `invenio-records-resources` and not `invenio-drafts-resources` could use this feature too without copy pasting code